File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2015/Y15D16.kt
package org.gristle.pdxpuzzles.advent.y2015

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.parsing.gvs

class Y15D16(input: String) : Day {

    // regex to extract items and amounts from a string
    private val pattern = Regex("""([a-z]+): (\d+)""")

    // takes a string and makes a "Sue." Items mapped to associated amounts
    private fun buildSue(s: String): Map<String, Int> = buildMap {
        s.gvs(pattern).forEach { (item, amt) -> put(item, amt.toInt()) }
    }

    // parses input
    private val sues: List<Map<String, Int>> = input
        .lineSequence()
        .map(::buildSue)
        .toList()

    // the gift-giving sue is defined by the ticker tape given in the puzzle
    private val auntSue: Map<String, Int>

    init {
        val tickerTape = """children: 3\n" +
            "cats: 7\n" +
            "samoyeds: 2\n" +
            "pomeranians: 3\n" +
            "akitas: 0\n" +
            "vizslas: 0\n" +
            "goldfish: 5\n" +
            "trees: 3\n" +
            "cars: 2\n" +
            "perfumes: 1"""
        auntSue = buildSue(tickerTape)
    }

    // returns true when for each item that the gift-giving Sue has, the Sue in question either doesn't have
    // an entry or has the same amount. Otherwise, returns false.
    private fun modernRetroencabulator(sue: Map<String, Int>): Boolean = auntSue.entries
        .all { (item, amt) -> (sue[item] ?: amt) == amt }

    // returns true when for each item that the gift-giving Sue has, the Sue in question either doesn't have
    // an entry or has the correct amount as defined in part 2 of the puzzle. Otherwise, returns false.
    private fun outdatedRetroencabulator(sue: Map<String, Int>): Boolean = auntSue.entries
        .all { (item, amt) ->
            if (sue.contains(item)) { // if Sue in question has an entry for the item, further analysis needed...
                val sueAmt = sue.getValue(item)
                when (item) {
                    "cats", "trees" -> sueAmt > amt
                    "pomeranians", "goldfish" -> sueAmt < amt
                    else -> sueAmt == amt
                }
            } else { // Sue in question has no entry, so move on to next item
                true
            }
        }

    override fun part1() = sues.indexOfFirst(::modernRetroencabulator) + 1

    override fun part2() = sues.indexOfFirst(::outdatedRetroencabulator) + 1
}

fun main() = Day.runDay(Y15D16::class)

//    Class creation: 39ms
//    Part 1: 40 (2ms)
//    Part 2: 241 (1ms)
//    Total time: 43ms

File: rust/advent/src/bin/2015/y15d16.rs
use advent::utilities::get_input::get_input;
use lazy_regex::regex;
use rustc_hash::FxHashMap;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input<'a> = (Sue<'a>, Vec<Sue<'a>>);
type Output = usize;
type Sue<'a> = FxHashMap<&'a str, usize>;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(15, 16).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn build_sue(s: &str) -> Sue<'_> {
    let rx = regex!(r"([a-z]+): (\d+)");
    rx.captures_iter(s)
        .map(|cap| {
            let item = cap.get(1).unwrap().as_str();
            let amt = cap.get(2).unwrap().as_str().parse().unwrap();
            (item, amt)
        })
        .collect()
}

fn parse_input(input: &str) -> Input<'_> {
    let ticker_tape = r"
        children: 3
        cats: 7
        samoyeds: 2
        pomeranians: 3
        akitas: 0
        vizslas: 0
        goldfish: 5
        trees: 3
        cars: 2
        perfumes: 1
    ";
    let aunt_sue = build_sue(ticker_tape);
    let sues = input.lines().map(build_sue).collect();
    (aunt_sue, sues)
}

fn solve<F>(aunt_sue: &Sue, sues: &Vec<Sue>, predicate: F) -> Output
where
    F: Fn(&str, usize, usize) -> bool,
{
    sues.iter()
        .position(|sue| {
            aunt_sue.iter().all(|(&item, &aunt_amt)| {
                sue.get(item)
                    .map(|&sue_amt| predicate(item, sue_amt, aunt_amt))
                    .unwrap_or(true)
            })
        })
        .unwrap() + 1
}

fn part1((aunt_sue, sues): &Input) -> Output {
    solve(&aunt_sue, &sues, |_, sue_amt, aunt_amt| sue_amt == aunt_amt)
}

fn part2((aunt_sue, sues): &Input) -> Output {
    solve(&aunt_sue, &sues, |item, sue_amt, aunt_amt| {
        match item {
            "cats" | "trees" => sue_amt > aunt_amt,
            "pomeranians" | "goldfish" => sue_amt < aunt_amt ,
            _ => sue_amt == aunt_amt,
        }
    })
}

#[test]
fn default() {
    let input = get_input(15, 16).unwrap();
    let input = parse_input(&input);
    assert_eq!(40, part1(&input));
    assert_eq!(241, part2(&input));
}

// Input parsed (657μs)
// 1. 40 (8μs)
// 2. 241 (15μs)
// Total: 682μs
